cluster.tf

```diff
@@ -1,104 +1,92 @@
+  vpc_id            = local.workload_cluster.vpc_id
+  vpc_subnets = {
+    vsi-zone-1 = [
+      for zone in local.workload_cluster.subnets :
+      {
+        id         = zone.id
+        zone       = zone.zone
+        cidr_block = zone.cidr
+      }
+    ]
   }
-
-  dynamic "kms_config" {
-    for_each = each.value.kms_config == null ? [] : [each.value.kms_config]
-    content {
-      crk_id           = module.key_management.key_map[kms_config.value.crk_name].key_id
-      instance_id      = module.key_management.key_management_guid
-      private_endpoint = kms_config.value.private_endpoint
-    }
+  worker_pools                       = var.worker_pools
+  ocp_version                        = var.ocp_version
+  cluster_tags                       = var.resource_tags
+  use_existing_cos                   = true
+  disable_public_endpoint            = local.workload_cluster.disable_public_endpoint
+  existing_cos_id                    = local.workload_cluster.cos_instance_crn
+  existing_key_protect_root_key_id   = module.key_management.key_map[local.workload_cluster.kms_config.crk_name].key_id
+  existing_key_protect_instance_guid = module.key_management.key_management_guid
+  logdna_instance_name               = module.observability_instances[local.workload_cluster.cluster_name].logdna_name
+  logdna_ingestion_key               = module.observability_instances[local.workload_cluster.cluster_name].logdna_ingestion_key
+  sysdig_instance_name               = module.observability_instances[local.workload_cluster.cluster_name].sysdig_name
+  sysdig_access_key                  = module.observability_instances[local.workload_cluster.cluster_name].sysdig_access_key
+  providers = {
+    helm = helm.workload_cluster
   }
-
-  disable_public_service_endpoint = true
-
-  timeouts {
-    create = "3h"
-    delete = "2h"
-    update = "3h"
-  }
-
 }
 
-##############################################################################
-
-
-##############################################################################
-# Create Worker Pools
-##############################################################################
-
-resource "ibm_container_vpc_worker_pool" "pool" {
-  for_each          = local.worker_pools_map
-  vpc_id            = each.value.vpc_id
-  resource_group_id = local.resource_groups[each.value.resource_group]
-  entitlement       = each.value.entitlement
-  cluster           = ibm_container_vpc_cluster.cluster[each.value.cluster_name].id
-  worker_pool_name  = each.value.name
-  flavor            = each.value.flavor
-  worker_count      = each.value.workers_per_subnet
+locals {
+  management_cluster = length(module.dynamic_values.clusters_map) == 2 ? module.dynamic_values.clusters_map["${var.prefix}-management-cluster"] : null
+}
 
-  dynamic "zones" {
-    for_each = each.value.subnets
-    content {
-      subnet_id = zones.value["id"]
-      name      = zones.value["zone"]
-    }
+module "management_cluster" {
```

Comment:
Instead of creating a new module, should you do a for_each on the cluster_map?  What if the user wants a cluster in the management vpc vs workload vpc?

Reply:
Initially, I tried using for_each on the cluster_map, but that creates a list of clusters as output and that list cannot be passed to the helm provider to configure the cluster.
a helm provider can be used only to configure a single cluster, looping the helm provider to configure the two clusters is not possible. We are using the helm provider to install the observability agents.

But, I have updated the logic to have a management cluster created without a workload cluster.

Reply:
For the observability, can we use this resource - https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/resources/ob_logging

Reply:
The observability instances are created in the observability.tf, but the logdna and the sysdig agents are installed on the cluster using the helm_release provider for which we had to separate out the creation of each cluster.